feat: Send and process ACK-ECN

larseggert opened the pull request:
Most of the remaining bits from #1495

Depends on #1604, will retarget this at mozilla/neqo once that has landed

larseggert committed:
feat: Send and process ACK-ECN

Most of the remaining bits from #1495

Depends on #1604

larseggert committed:
Enable ECN interop test

larseggert committed:
Implement ECN validation

mxinden committed:
fix(http3): typos (#1669)

mxinden committed:
refactor: remove dead fairness and sendorder fn on SendStream (#1668)

Removes `fn` `set_sendorder` and `set_fairness` on `SendStream` trait.

The `SendStream` trait is not publicly exposed. Neither `set_sendorder` nor
`set_fairness` is called within `neqo-http3`.

Co-authored-by: Martin Thomson <mt@lowentropy.net>

mxinden commented on neqo-transport/src/connection/mod.rs:
```diff
@@ -2902,6 +2920,44 @@ impl Connection {
+                    "ACK had {} new marks, but sent {} packets, disabling ECN",
+                    ecn_diff[cur_mark],
+                    newly_acked
+                );
```
Based on the enclosing `if` we know that ECN is enabled on the path. `ecn_diff[cur_mark]` ignores all CE marks. Wouldn't the following be correct?

```suggestion
                qwarn!(
                    "ACK had {} new marks, but acked {} packets, disabling ECN",
                    sum_inc,
                    newly_acked
                );
```

mxinden commented on neqo-transport/src/tracking.rs:
```diff
@@ -397,9 +418,15 @@ impl RecvdPackets {
         }
     }
 
+    /// Increase the ECN count for the mark given by `ecn` by `n`.
+    pub fn inc_ecn_count(&mut self, ecn: IpTosEcn, n: u64) {
+        self.ecn_count[ecn] += n;
+    }
```
Nit, given that `inc_ecn_count` is only ever called with `n = 1`:

```suggestion
    /// Increase the ECN count for the mark given by `ecn` by `1`.
    pub fn inc_ecn_count(&mut self, ecn: IpTosEcn) {
        self.ecn_count[ecn] += 1;
    }
```

larseggert replied:
I had that originally, but was wondering if with multi-packet I/O we might want to increase by more than one. But I guess we can make that change at that point.

mxinden commented on neqo-transport/src/tracking.rs:
```diff
@@ -16,18 +16,20 @@ use std::{
     stats::FrameStats,
 };
 
 // TODO(mt) look at enabling EnumMap for this: https://stackoverflow.com/a/44905797/1375574
```
With the addition below (`#[derive(..., Enum)]`) is this comment still necessary?

larseggert replied:
I think @martinthomson meant something other by the "this" in his comment.

mxinden commented on neqo-client/src/main.rs:
```diff
@@ -241,8 +241,7 @@ impl Args {
         match testcase.as_str() {
-            // TODO: Add "ecn" when that is ready.
-            "http3" => {}
+            "http3" | "ecn" => {}
```
Missing corresponding change in `server/src/main.rs`:

``` diff
modified   neqo-server/src/main.rs
@@ -734,7 +734,7 @@ async fn main() -> Result<(), io::Error> {
         }
 
         match testcase.as_str() {
-            "http3" => (),
+            "http3" | "ecn" => (),
             "zerortt" => {
                 args.use_old_http = true;
                 args.alpn = String::from(HQ_INTEROP);
```

mxinden reviewed (commented):
I have rebased the commits onto latest https://github.com/mozilla/neqo/pull/1604 and added a couple of smaller patches.

I have run this through the ECN QUIC Interop testcase. As of now, the server receives ECT0 from the client, but the server fails to properly send ECT0 to the client. This is due to wrong ECN handling of IPv4-Mapped IPv6 addresses. I will have a patch coming shortly. With the patch the ECN testcase between a neqo client and server succeeds.

//CC @larseggert to prevent duplicate work.

mxinden committed:
chore: remove neqo-interop (#1672)

`neqo-interop` is replaced by `neqo-client` and `neqo-server` and thus unused.

mxinden committed:
feat: update to Rust 2021 edition (#1671)

* feat: update to Rust 2021 edition

See transition docs:

https://doc.rust-lang.org/edition-guide/editions/transitioning-an-existing-project-to-a-new-edition.html

The described `cargo fix --edition` does not yield any suggestions.

* fix: clippy

---------

Co-authored-by: Martin Thomson <mt@lowentropy.net>

martinthomson committed:
Extra clippy (#1664)

* Add some more doc-stuff

* Get rid of redundant imports

No more use of `use super::*` in tests.
Avoid importing things that are in the rust prelude.

mxinden committed:
refactor: derive default for unit enum variants (#1667)

For unit enum default variants, derive `Default` and use the `#[default]` attribute.

See https://doc.rust-lang.org/std/default/trait.Default.html#enums for details.

Co-authored-by: Martin Thomson <mt@lowentropy.net>

mxinden commented:
> I have run this through the ECN QUIC Interop testcase. As of now, the server receives ECT0 from the client, but the server fails to properly send ECT0 to the client. This is due to wrong ECN handling of IPv4-Mapped IPv6 addresses. I will have a patch coming shortly. With the patch the ECN testcase between a neqo client and server succeeds.

Fixed with https://github.com/quinn-rs/quinn/pull/1765. Updated in https://github.com/mozilla/neqo/pull/1604 via https://github.com/mozilla/neqo/pull/1604/commits/a0d309385bfe6510593e74591f4f683ce0dd6ced.

mxinden commented:
For convenience, I pushed my changes to https://github.com/mxinden/neqo/tree/feat-more-ecn-2. It includes:

- A rebase onto latest https://github.com/mozilla/neqo/pull/1604, including https://github.com/quinn-rs/quinn/pull/1765.
- A commit for each of the suggestions above.

mxinden committed:
feat(.github): add merge_group trigger to workflows (#1674)

Enables workflows to run on pull requests within the merge queue.

martinthomson committed:
Benchmark tweaks (#1663)

* Benchmark tweaks

1. Use NSS Release
2. Add workflow_dispatch
3. Move some env values up in the file

* YAML corrections

Co-authored-by: Lars Eggert <lars@eggert.org>
Signed-off-by: Martin Thomson <mt@lowentropy.net>

* Build NSS optimized

---------

Signed-off-by: Martin Thomson <mt@lowentropy.net>
Co-authored-by: Lars Eggert <lars@eggert.org>

mxinden committed:
refactor(client,server): use quinn_udp for I/O (#1604)

* Initial commit

* refactor(server): replace mio with tokio

* Move ready logic into fn

* Extend expect docs

* Restrict tokio features

* Only process datagram once

* Remove superfluous pub

* fmt

* Fix send path

* Fix receive path

* Instantiate socket state once

* Fix busy loop

* Have neqo-client use quinn-udp

* Add TODO

* Await writable

* Unify tx and rx

* Introduce wrapper type Socket

* Move bind to common

* Check if datagram was sent as a whole and avoid allocation

* Make into_data pub(crate)

* Refactor send

* Reference issue

* Remove debugs

* Fix test

* Reduce diff

* Reduce diff

* Pin quinn-udp to rev

* Address clippy lints

* fmt

* fmt

* clippy

* clippy

* Pass None ttl

Not yet supported by quinn-udp.

* Debug race condition on Windows

* Debug windows failure

* Have receiver use random port

* Test with Ect1 instead of Ce

Windows does not allow setting Ce:

> Your application isn't allowed to specify the Congestion Encountered (CE) code point when sending datagrams. The send will return with error WSAEINVAL.

https://learn.microsoft.com/en-us/windows/win32/winsock/winsock-ecn

* Revert "Debug windows failure"

This reverts commit e9ac36cb557a2edd5be3654b37202199ca4634ef.

* Revert "Debug race condition on Windows"

This reverts commit 6f330d353eabaae5615d24aa3a7bcbd1549e2f26.

* Fold tos_tx

* Add reason to clippy lint ignore

* fix: include quinn-udp IPv4-mapped IPv6 patch

https://github.com/quinn-rs/quinn/pull/1765

---------

Co-authored-by: Lars Eggert <lars@eggert.org>

larseggert committed:
Update neqo-transport/src/connection/mod.rs

Co-authored-by: Max Inden <mail@max-inden.de>

larseggert committed:
Update neqo-transport/src/tracking.rs

Co-authored-by: Max Inden <mail@max-inden.de>

larseggert committed:
Address code review comment

larseggert committed:
Merge branch 'feat-more-ecn-2' into feat-more-ecn

larseggert committed:
Merge branch 'mozilla:main' into feat-more-ecn

larseggert committed:
Fixes

larseggert commented:
See https://github.com/mozilla/neqo/pull/1678
